### PR TITLE
[MXNET-918] [Introduce Random module / Refact code generation (#13038)][Cherry pick] 

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##
Cherry pick @mdespriee change to Java API. Applied minor change in order to make it compatible with Java API Generation

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by zachgk** [comment]
@mxnet-label-bot add [Java]

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 1]
```diff
@@ -17,196 +17,147 @@
...
+        else ele._2.last
+      }
+    }).map(absClassFunction => {
         generateJavaAPISignature(absClassFunction)
-      })
+      }).toSeq
     val packageName = "NDArrayBase"
     val packageDef = "package org.apache.mxnet.javaapi"
-    fileGen(filePath, packageName, packageDef, absFuncs)
+    writeFile(filePath + "javaapi/", packageName, packageDef, absFuncs)
```
Comment (lanking520) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: 95-109 changed

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 2]
```diff
@@ -17,196 +17,147 @@
...
 
-  def generateJavaAPISignature(func : absClassFunction) : String = {
+  def generateJavaAPISignature(func : Func) : String = {
     val useParamObject = func.listOfArgs.count(arg => arg.isOptional) >= 2
     var argDef = ListBuffer[String]()
     var classDef = ListBuffer[String]()
     var requiredParam = ListBuffer[String]()
     func.listOfArgs.foreach(absClassArg => {
-      val currArgName = safetyNameCheck(absClassArg.argName)
+      val currArgName = absClassArg.safeArgName
```
Comment (lanking520) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: 154-160

File: scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/JavaNDArrayMacro.scala [hunk 3]
```diff
@@ -17,8 +17,7 @@
```
Comment (lanking520) [inline scala-package/macros/src/main/scala/org/apache/mxnet/javaapi/JavaNDArrayMacro.scala]: Change applied on this file

File: scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala [hunk 4]
```diff
@@ -17,196 +17,147 @@
...
     val notGenerated = Set("Custom")
-    val absClassFunctions = getSymbolNDArrayMethods(false, true)
-    // TODO: Add Filter to the same location in case of refactor
-    val absFuncs = absClassFunctions.filterNot(_.name.startsWith("_"))
-      .filterNot(ele => notGenerated.contains(ele.name))
-      .map(absClassFunction => {
+    val absClassFunctions = functionsToGenerate(false, false, true)
+    val absFuncs = absClassFunctions.filterNot(ele => notGenerated.contains(ele.name))
+      .groupBy(_.name.toLowerCase).map(ele => {
+      // Pattern matching for not generating depreciated method
```
Comment (andrewfayres) [inline scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: I'm not crazy about this logic but based on how you've explained the problem to me I don't really know what else we can do here. Please add more detail in this comment explaining that we sometimes get two methods with the same name but capitalization differences and the only existing mechanism for to determine which one has been deprecated is the capitalization.

Comment (andrewfayres) [reply scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala]: please also copy the comment to the corresponding code in the JavaNDArrayMacro

**Pull request merged**